models/correlospinmatrix_detector.cpp

```diff
@@ -0,0 +1,433 @@
+  , tau_max_( 10 * delta_tau_ )
+  , Tstart_( Time::ms( 0.0 ) )
+  , Tstop_( Time::pos_inf() )
+  , N_channels_( 1 )
+{
+}
+
+nest::correlospinmatrix_detector::Parameters_::Parameters_( const Parameters_& p )
+  : delta_tau_( p.delta_tau_ )
+  , tau_max_( p.tau_max_ )
+  , Tstart_( p.Tstart_ )
+  , Tstop_( p.Tstop_ )
+  , N_channels_( p.N_channels_ )
+{
+  // Check for proper properties is not done here but in the
+  // correlospinmatrix_detector() copy c'tor. The check cannot be
+  // placed here, since this c'tor is also used to copy to
+  // temporaries in correlospinmatrix_detector::set_status().
+  // If we checked for errors here, we could never change values
+  // that have become invalid after a resolution change.
+  delta_tau_.calibrate();
+  tau_max_.calibrate();
+  Tstart_.calibrate();
+  Tstop_.calibrate();
+}
+
+nest::correlospinmatrix_detector::State_::State_()
+  : n_events_( 1, 0 )
+  , incoming_()
+  , last_i_( 0 )
+  , t_last_in_spike_( Time::neg_inf() )
+  , count_covariance_( 1, std::vector< std::vector< long_t > >( 1, std::vector< long_t >() ) )
+{
+}
+
+
+/* ----------------------------------------------------------------
+ * Parameter extraction and manipulation functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::correlospinmatrix_detector::Parameters_::get( DictionaryDatum& d ) const
+{
+  ( *d )[ names::delta_tau ] = delta_tau_.get_ms();
+  ( *d )[ names::tau_max ] = tau_max_.get_ms();
+  ( *d )[ names::Tstart ] = Tstart_.get_ms();
+  ( *d )[ names::Tstop ] = Tstop_.get_ms();
+  ( *d )[ names::N_channels ] = N_channels_;
+}
+
+void
+nest::correlospinmatrix_detector::State_::get( DictionaryDatum& d ) const
+{
+  ( *d )[ names::n_events ] = IntVectorDatum( new std::vector< long_t >( n_events_ ) );
+
+  ArrayDatum* CountC = new ArrayDatum;
+  for ( size_t i = 0; i < count_covariance_.size(); ++i )
+  {
+    ArrayDatum* CountC_i = new ArrayDatum;
+    for ( size_t j = 0; j < count_covariance_[ i ].size(); ++j )
+    {
+      CountC_i->push_back(
+        new IntVectorDatum( new std::vector< long_t >( count_covariance_[ i ][ j ] ) ) );
+    }
+    CountC->push_back( *CountC_i );
+  }
+  ( *d )[ names::count_covariance ] = CountC;
+}
+
+bool
+nest::correlospinmatrix_detector::Parameters_::set( const DictionaryDatum& d,
+  const correlospinmatrix_detector& n )
+{
+  bool reset = false;
+  double_t t;
+  long_t N;
+
+  if ( updateValue< long_t >( d, names::N_channels, N ) )
+  {
+    if ( N < 1 )
+    {
+      throw BadProperty( "/N_channels can only be larger than zero." );
+    }
+    else
+    {
+      N_channels_ = N;
+      reset = true;
+    }
+  }
+
+  if ( updateValue< double_t >( d, names::delta_tau, t ) )
+  {
+    delta_tau_ = Time::ms( t );
+    reset = true;
+  }
+
+  if ( updateValue< double_t >( d, names::tau_max, t ) )
+  {
+    tau_max_ = Time::ms( t );
+    reset = true;
+  }
+
+  if ( updateValue< double_t >( d, names::Tstart, t ) )
```

Comment:
do we want Tstart and Tstop to be non-negative and also Tstop > Tstart? if so we need checks here.